cirq-core/cirq/ops/control_values.py

```diff
@@ -0,0 +1,272 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Collection, Optional, Sequence, Union, Tuple, List, Type, cast, TYPE_CHECKING
+
+import copy
+import itertools
+
+if TYPE_CHECKING:
+    import cirq
+
+
+def flatten(sequence):
+    def _flatten_aux(sequence):
+        if isinstance(sequence, int):
+            yield sequence
+        else:
+            for item in sequence:
+                yield from _flatten_aux(item)
+
+    return tuple(_flatten_aux(sequence))
+
+
+def _from_int(val: int) -> Tuple[Tuple[int, ...], ...]:
+    return ((val,),)
+
+
+def _from_sequence_int(vals: Sequence[int]) -> Tuple[Tuple[int, ...], ...]:
+    return tuple((val,) for val in vals)
+
+
+def _from_sequence_sequence(vals: Sequence[Sequence[int]]) -> Tuple[Tuple[int, ...], ...]:
+    return tuple(tuple(product) for product in vals)
+
+
+class ControlValues:
+    def __init__(
+        self, control_values: Sequence[Union[int, Collection[int], Type['ControlValues']]]
+    ):
+        if len(control_values) == 0:
+            self._vals = cast(Tuple[Tuple[int, ...], ...], (()))
+            self._num_variables = 0
+            self._nxt = None
+            return
+        self._nxt = None
+
+        if len(control_values) > 1:
+            self._nxt = ControlValues(control_values[1:])
+
+        if isinstance(control_values[0], ControlValues):
+            aux = control_values[0].copy()
+            aux.product(self._nxt)
+            self._vals, self._num_variables, self._nxt = aux._vals, aux._num_variables, aux._nxt
+            self._vals = cast(Tuple[Tuple[int, ...], ...], self._vals)
+            return
+
+        val = control_values[0]
+        if isinstance(val, int):
+            self._vals = _from_int(val)
+        elif isinstance(val, (list, tuple)):
+            if isinstance(val[0], int):
+                self._vals = _from_sequence_int(val)
+            else:
+                self._vals = _from_sequence_sequence(val)
+        else:
+            raise TypeError(f'{val} is of Unsupported type {type(val)}')
+        self._num_variables = len(self._vals[0])
+
+    def product(self, other):
+        """Sets self to be the cartesian product of all combinations in self x other.
+
+        Args:
+            other: A ControlValues object
+        """
+        if other is None:
+            return
+        other = other.copy()
+        cur = self
+        while cur._nxt is not None:
+            cur = cur._nxt
+        cur._nxt = other
+
+    def __call__(self):
+        return self.__iter__()
+
+    def __iter__(self):
+        nxt = self._nxt if self._nxt else lambda: [()]
+        if self._num_variables:
+            return itertools.product(self._vals, nxt())
+        else:
+            return itertools.product(*(), nxt())
+
+    def copy(self):
+        """Returns a deep copy of the object."""
```

Comment:
Once everything has been implemented functionally, there should be no need for this function.

Reply:
done